posthog/hogql_queries/utils/query_date_range.py

```diff
@@ -55,11 +55,12 @@ def date_to(self) -> datetime:
                 now=self.now_with_timezone,
             )
 
-        is_relative = not self._date_range or not self._date_range.date_to or delta_mapping is not None
-        if not self.is_hourly:
-            date_to = date_to.replace(hour=23, minute=59, second=59, microsecond=999999)
-        elif is_relative:
-            date_to = date_to.replace(minute=59, second=59, microsecond=999999)
+        if not self._date_range or not self._date_range.explicitDate:
```

Comment:
probs need similar handling in date_from? Since that can be explicit too, and this should apply to both dates :)

Reply:
sorry, I'm a bit confused, haven't been following the hogql query updates too closely.

The link you're pointing to is the equivalent filter object for legacy insights right? Which supports explicit_date for date_from & date_to ( https://github.com/PostHog/posthog/blob/master/posthog/queries/query_date_range.py#L65 )

But, this is the component for hogql queries, and is independent of the filter one (is it not?) - and Line 80 below always truncates date_from to start of date, and same in Line 83 with the `.replace()` - so right now (I expect) date_from is always truncated in hogql queries irrespective of whats passed in.

I think just having another test here: https://github.com/PostHog/posthog/pull/21305/files#diff-ba7a04fe95294245e7cb93bb4cd8d621f4e6e68c8390457320da4fe2496938b5R107-R116 like with date_to for date_from would make me a lot more confident of this 😅

Reply:
I'm glad your keeping an eye on this :)

> The link you're pointing to is the equivalent filter object for legacy insights right? Which supports explicit_date for date_from & date_to ( https://github.com/PostHog/posthog/blob/master/posthog/queries/query_date_range.py#L65 )

Yep, that's right.

> But, this is the component for hogql queries, and is independent of the filter one (is it not?) - and Line 80 below always truncates date_from to start of date, and same in Line 83 with the .replace() - so right now (I expect) date_from is always truncated in hogql queries irrespective of whats passed in.

It does, but so does the legacy version. The line where the legacy version changes its behaviour based on the `explicit_date` flag doesn't exist in the HogQL version (I don't know why this is). See the following diff: Left is HogQL and right is legacy. The lines based on `explicit_date` are line 14 and 15 and they are missing from the HogQL version.
<img width="1512" alt="Screenshot 2024-04-04 at 00 41 21" src="https://github.com/PostHog/posthog/assets/1851359/3871a2ed-16db-44b2-bbdf-e39471907c78">

> I think just having another test here: https://github.com/PostHog/posthog/pull/21305/files#diff-ba7a04fe95294245e7cb93bb4cd8d621f4e6e68c8390457320da4fe2496938b5R107-R116 like with date_to for date_from would make me a lot more confident of this 😅

The test already tests for `date_to` and `date_from`. I have checked one the generated output for one query as well and there the `date_from` was already ok, just the `date_to` changes with this PR.

Reply:
ok I'm blind, just seeing the test checks both 🐵 🙈  - thanks for explaining!

Oh interesting, does this mean HogQL always uses the explicit date passed in for date-from, unless its a relative date? huh

Reply:
> Oh interesting, does this mean HogQL always uses the explicit date passed in for date-from, unless its a relative date? huh

Yep.